File: orchalang-orchacompiler/src/main/java/transactionJpa/PopulateDatabase.java
package transactionJpa;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.sql.DataSource;
import java.util.List;

@Component
public class PopulateDatabase {


    @Autowired
    private EntityManagerFactory entityManagerFactory;

    @Autowired
    private DataSource dataSource;

    @Transactional
    public void saveStudent(StudentDomain student) throws Exception {
        System.out.print("saveStudent receives: " + student);
        EntityManager entityManager = entityManagerFactory.createEntityManager();
        entityManager.joinTransaction();
        entityManager.merge(student);
        System.out.println(" and save it.");
    }


    public List<?> readDatabase(){
        JdbcTemplate jdbcTemplate = new JdbcTemplate(this.dataSource);
        List<?> results = jdbcTemplate.queryForList("Select * from Student");
        return results;
    }

}

File: orchalang-orchacompiler/src/main/kotlin/school/PopulateDatabase.kt
package school

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManagerFactory
import javax.sql.DataSource

@Component
class PopulateDatabase {
    @Autowired
    private val entityManagerFactory: EntityManagerFactory? = null

    @Autowired
    private val dataSource: DataSource? = null

    @Transactional
    @Throws(Exception::class)
    fun saveStudent(student: StudentDomain) {
        print("ici " + entityManagerFactory)
        print("saveStudent receives: $student")
        val entityManager = entityManagerFactory!!.createEntityManager()
        entityManager.joinTransaction()
        entityManager.merge<Any>(student)
        println(" and save it.")
    }

    @Throws(Exception::class)
    fun enrollStudent(student: StudentDomain): StudentDomain {
        print("enrollStudent receives: $student")
        val entityManager = entityManagerFactory!!.createEntityManager()
        entityManager.joinTransaction()
        val studentDomain = StudentDomain("Ben", 25, 5)
        entityManager.persist(studentDomain) // attention même si merge est utilisé, si nouvelle tx => nouvelle insertion !
        println(" and returns: $student")
        return student
    }

    fun readDatabase(): List<*> {
        val jdbcTemplate = JdbcTemplate(dataSource!!)
        return jdbcTemplate.queryForList("Select * from Student")
    }
}